### DISCUSSION TITLE
how to search for a value in two or more tables?

### DISCUSSION THREAD

**Discussion by saman2007** [supabase/supabase#7929] [Questions]
hello. i hope you feel good.
im creating a simple music platform like spotify. i have playlists, musics and artists tables in supabase. i have a search bar that user can search a value. the results of a searched value can be an artist, music or playlist. so i should search for the searched value in musics, playlists and artists tables. but i dont know how. i can send 3 requests to these 3 tables but its not ok because it takes more time to return the results from database.  and i want to return full row if the value of a specific column is equal or like the searched value.
for example imagine these rows in different tables:

artists:
name|musics|genres|
-------|--------|-------|
eminem|200|["rap", "hip hop"]|

playlists:
name|musics|genres|musicsObj|
------|-------|--------|-----------|
a playlist for you|102|["rap", "hip hop", "pop"]|[{musicInfos}]

musics:
name|singer|genres|cover|
------|-------|-------|------|
without me|eminem|["rap"]|cover image link

user search text: rap musics ==> results: all the above rows
user search: eminem ==> results: the first row and last row
user search: without me ===> results: the lats row
user search: without me-eminem ==> results: the last row

how can i implement something like this?
thanks for helping.

**Comment by wcooperscott** [comment]
Option:
Please read [filtering with inner joins](https://supabase.com/docs/reference/javascript/select#filtering-with-inner-joins) for SupabaseJS. I have never tried to join more than two tables together so I do not know if it will work for 3+ but it's worth a try. I would start with 2 (to prove you understand how to use the inner() function) and then attach a 3rd join. If you get this working, please let me know. If you need other join types, you will need to use one of the other options below.

Option:
Create a SQL function. You can execute the function from your client by calling `supabase.rpc(name, params)`.

You can read about Supabase SQL functions here: https://supabase.com/docs/guides/database/functions

The contents of your SQL function `search_songs` might look something like:
```
SELECT *
FROM musics
INNER JOIN artists
ON musics.id = artists.musics
INNER JOIN playlists
ON musics.id = playlists.musics
```

SQL functions are executed with the row level security permissions of the user that executes them, so make sure the user has ability to select rows from all 3 tables.

That function body example would, of course, return every song. Be certain to apply a limit and paginate by cursor so you don't query thousands of rows if table gets large. This is nice because you can also attach additional filters to the SQL function from the client. For example:
```
let query = supabase.rpc('search_songs')
   .ilike('name', search_text)
   .order('created_at', { ascending: true })
   .limit(20)
if(cursor)
   query.lt('created_at', cursor)

const searchResults = await query;
```
where `cursor` is a `date.toISOString()` and only provided if it is retrieving more data. You can read about SQL pagination [here](https://dev.to/appwrite/this-is-why-you-should-use-cursor-pagination-4nh5). In this example, I use `cursor pagination` by `date` (created_at).


Option:
Another option is to create a SQL `view`. A view is _like_ a table but it is generated by select statements. Your view creation code would be very similar to that of the SQL sample code I provided above. The downside of a view is that you lose the ability to apply row level security. And the limit/pagination logic should also be applied when selecting from the view.

**Reply by saman2007** [reply]
thanks for your answer. since im a frontend developer, its really hard for me to work with postgreSQL. because it has a lot of keywords and etc. i tried a lot to create a function and return the rows, but i couldnt. i thought i could do something like this by using the methods that supabase js library gives us. cant i do this by using the methods that supabase library gives us?

**Reply by wcooperscott** [reply]
@saman2007 I've updated the post to include another option. You may find it a favorable solution.

**Reply by saman2007** [reply]
thank you for your great answer and your time.
to use filtering with inner joins, i should add foreign key relationship. i added it and i understood that it only gives me one row that is exactly like the value of the relation column. for example if i add foreign relation from id column in artists table to id column in musics table, i get only the row from artists that its id is the same with id of the row in musics table. i think you got my point. so this is useless for my situation.
about creating functions in postgreSQL, i read that link you sent me and i searched a lot about it. actually i know how to create function in postgreSQL but i dont know what type should i return from the function with the code you sent in your answer. i said the other problem in the code.

```
create
or replace function hello_world() -- 1
returns /* wehat should i return here? what type? */
language sql as $$
  /*
  * here i actually want to select all the rows of table artists, musics and playlists.
  * and i can filter the rows by using methods of supabase js library.
  * i dont want to return a row that for example its column id is equal to column id in an other table. i think you got my point
  */
$$;
```

again, thanks for your time. i think im so close to the solution if you help me :)

**Reply by wcooperscott** [reply]
Did you find a solution? I might have time to write you a response, tonight

**Reply by saman2007** [reply]
No. I couldnt find yet...
